### ISSUE TITLE
What are the factors affecting the accuracy of YOLOv11 Pose Keypoint Detection?

### ISSUE THREAD

**Issue by ipmLessing** [ultralytics/ultralytics#19284]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

# What are the factors affecting the accuracy of YOLOv11 Pose Keypoint Detection?

### Training Details
- **Image Size:** 640 x 360  
- **Dataset Size:** 200K images (bbox + keypoints)  
- **Model:** YOLOv11 Pose  

### Tuning Attempts
I have tried all possible tuning parameters, including but not limited to:
- `rect`, `single_cls`, `multi_scale`, `cos_lr`, `augmentation`
- `box`, `cls`, `pose`, `kobj`, `nbs`, etc.  

The best result so far is with:
- `augment=True`, `degrees=180`, `shear=5`, `perspective=0.0001`

### Issues Observed
1. **General Inaccuracy:**  
   The predicted keypoints are consistently shifted, regardless of the model size or configuration.
2. **Rotation Artifacts:**  
   When rotation is applied, the keypoints follow the rotation of the image but are noticeably shifted, leading to very strange results.  

### Additional Attempts
- I have tried modifying the heads and backbone using different default modules, but the results are still unsatisfactory.

### Comparison with Other Models
As a comparison, I built a CNN-based Gaussian heatmap network (e.g., MobileNet backbone + heatmap head) with the same amount of data. The results are significantly better than YOLOv11 Pose.  

The only reason I'm persisting with YOLOv11 is its speed, which is crucial for deployment on edge devices.

---

### Possible Reasons for Keypoint Shifting (According to GPT)
1. **Anchor Misalignment:**  
   YOLO's anchor-based detection might not align well with the keypoint locations, especially during transformations like rotation.
2. **Interpolation Artifacts:**  
   When rotating or transforming images, interpolation can cause shifts in pixel values, affecting keypoint predictions.
3. **Loss Function Limitations:**  
   The current loss function might not penalize keypoint misalignment effectively.
4. **Pose-Specific Augmentation Sensitivity:**  
   Some augmentations (e.g., extreme rotations) could introduce noise or spatial inconsistencies, leading to shifting keypoints.
5. **Feature Map Resolution:**  
   Lower resolution feature maps might not capture fine-grained details necessary for accurate keypoint localization.
6. **Model Architecture Constraints:**  
   YOLO's architecture, optimized for object detection, might not be as effective for precise keypoint localization compared to heatmap-based approaches.
7. **Dataset Imbalance or Noise:**  
   If the dataset has biases, inconsistencies, or noisy labels, the model may struggle to generalize well.

---

### Request for Help
I am seeking insights or suggestions on:
- Why the keypoints might be consistently shifted (regression head v.s. heatmap head)
- How to improve the keypoint accuracy, especially when rotation is applied
- Any architectural changes or tuning parameters I might have overlooked

Thank you!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ipmLessing, thank you for your detailed question and for sharing your YOLOv11 Pose tuning experiences 🚀! It looks like you've done an impressive amount of groundwork here, and your observations and hypotheses are very insightful. 🌟 

We recommend referring to the [Ultralytics Docs](https://docs.ultralytics.com/) for comprehensive information, and the [Model Training Tips](https://docs.ultralytics.com/guides/model-training-tips/) guide might be especially helpful in addressing some of the challenges you're facing.

To assist further:  
- If this is a 🐛 Bug Report (e.g., shifted keypoints being incorrect or model failing under certain conditions), please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) along with explicit details on the training configuration and dataset setup. This would help us better debug and identify potential issues.  
- If your question pertains to improving accuracy ❓, please ensure you've provided detailed information about your dataset, augmentation strategies, and any logs (if available) related to training and validation. This will help guide us in offering more specific suggestions.  

### Suggestions

In the meantime, please consider the following:  
1. Upgrade to the latest version of the `ultralytics` package to ensure you're leveraging the newest features and bug fixes. Use the command:
   ```bash
   pip install -U ultralytics
   ```
2. Explore potential architectural tweaks, as you’ve already started. Keep in mind that YOLO's architecture is generally optimized for detection; feature-specific adjustments for keypoint tasks may yield improvements.
3. Validate the diversity and label quality of your dataset to mitigate any annotation or transformation biases.

### Explore Community Support

For further insights, we encourage you to join the Ultralytics community:  
- Real-time discussions? Join our [Discord](https://discord.com/invite/ultralytics) 🎧  
- In-depth Q&A? Visit our [Discourse](https://community.ultralytics.com/) forum.  
- Share knowledge? Check out the [Subreddit](https://reddit.com/r/Ultralytics).  

### Verified Environments  

Ensure you're testing your workflows in a supported and up-to-date environment. YOLO can be run in the following verified setups (with dependencies like [CUDA](https://developer.nvidia.com/cuda-zone), [Python](https://www.python.org/), and [PyTorch](https://pytorch.org/) preinstalled):  
- **Notebooks with free GPU**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>  
- **Cloud Deployments**: Consider [Google Cloud](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) or [AWS](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/).  
- **Docker**: Explore YOLO's [Docker image](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) for a reproducible environment <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>.  

### Status  

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>  
If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing, verifying the correct operation of various YOLO tasks.  

This is an automated response to ensure your query is acknowledged promptly. An Ultralytics engineer will review your issue and provide further assistance soon. 😊

**Comment by Y-T-G** [comment]
Do you have example images of shift?

**Comment by ipmLessing** [comment]
<img width="793" alt="Image" src="https://github.com/user-attachments/assets/f7eb95d0-6e38-4f31-b3cb-4c9165af89c1" />

I illustrated my case using a square as the object, with its four corners labeled as keypoints [1].  

After performing a grid search for hyperparameters, I noticed a consistent issue across all model sizes (from n to xl): the predictions were consistently shifted, as shown in [2].  

In general, the predicted keypoints tend to fall within a rotated region, as depicted in [3], even when the object itself is at a 0-degree orientation. The keypoints aren’t entirely incorrect or random, but when I include augmentation with a 180-degree rotation, the shift in the predictions becomes even more.

**Comment by Y-T-G** [comment]
The keypoints in YOLO Pose have specific meaning, so it can't be arbitrarily assigned. Each keypoint is a class of its own and should be both visually and semantically distinct from the others. The order of the keypoints are important and non-arbritrary. The model tries to learn the distinction between each keypoint.

So if you're assigning keypoints arbitrarily, without any meaning to the order, then the model would not be able to learn. Such as assigning keypoints to corners in your illustration where with random rotation, the model can no longer understand or learn what the difference between each keypoint is because it's inconsistent.

**Comment by ipmLessing** [comment]
@Y-T-G 

Thank you very much for your detailed explanation! I greatly appreciate it.

I was hoping to ask for a bit more clarification regarding the keypoints in YOLO Pose. You mentioned that they have specific meanings—could you kindly point me to any documentation where I can learn more about this?

To further clarify my question, in the yolo11-pose.yaml configuration file, I noticed the following:

nc=80 (number of classes)
kp_shape=(17, 3)

In my case, for example, I only have one class (nc=1, for a square) and kp_shape=(4, 3), representing the four corners. 
<img width="254" alt="Image" src="https://github.com/user-attachments/assets/28b187bf-aa0f-4180-b4ea-c18dae69b5d4" />
I would appreciate your guidance on the following points:

1. Is it possible to change nc? Does it have any specific meaning beyond the number of classes?
2. Could you explain the significance of each keypoint? Specifically, I am curious about the order, connections, and relationships among the 17 keypoints.
3. Is it possible to modify the number of keypoints? For instance, in my scenario, I only need four keypoints connected in the order A → B → C → D.
4. If the number of keypoints cannot be changed, what would be the recommended approach to adapt YOLO Pose for this custom case?

Thank you once again for your time and help! I truly appreciate your insights:)

**Comment by Y-T-G** [comment]
After rotation kp4 is now in the location of kp1. It's inconsistent and the model can't learn due to this inconsistency. It needs to be kp1 even after rotation, so that it's consistent, and the model learns that kp1 refers to the top-left corner.

![pose-rotation.png](https://github.com/user-attachments/assets/c263e28f-782a-4bfd-9071-6edee18c05c9)

Each keypoint needs to have a special meaning such as kp1 refers to top-left corner, kp2 refers to top-right and so on. And this meaning needs to be consistent across all images. The model wouldn't be able to learn if they are inconsistent because for YOLO Pose, the order of keypoints is important.

**Comment by Y-T-G** [comment]
> @Y-T-G 
> 
> Thank you very much for your detailed explanation! I greatly appreciate it.
> 
> I was hoping to ask for a bit more clarification regarding the keypoints in YOLO Pose. You mentioned that they have specific meanings—could you kindly point me to any documentation where I can learn more about this?

I don't know if it's documented. But it's how YOLO Pose is designed. The order is important. Otherwise, it wouldn't be able to differentiate between left-eye vs. right-eye, etc. It's learning to differentiate between all the keypoints. And this is also why `flip_idx` exists. Because if you horizontally flip the image of a person, left-eye becomes right-eye, and if you don't correct it, it will be inconsistent. It's corrected internally after flipping based on `flip_idx`.

> To further clarify my question, in the yolo11-pose.yaml configuration file, I noticed the following:
> 
> nc=80 (number of classes)
> kp_shape=(17, 3)
> 
> In my case, for example, I only have one class (nc=1, for a square) and kp_shape=(4, 3), representing the four corners. 
> <img width="254" alt="Image" src="https://github.com/user-attachments/assets/28b187bf-aa0f-4180-b4ea-c18dae69b5d4" />
> I would appreciate your guidance on the following points:
> 
> 1. Is it possible to change nc? Does it have any specific meaning beyond the number of classes?

You can change `nc`.

> 2. Could you explain the significance of each keypoint? Specifically, I am curious about the order, connections, and relationships among the 17 keypoints.

All of the 17 keypoints are based on COCO body keypoints. They all have specific meaning. Your data doesn't have to follow COCO body pose. But each keypoint should have a specific meaning. Like the first keypoint in your label represents top-left corner. So it should always be top-left, in all your labels. Otherwise, it would be inconsistent. With rotation/flip augmentations, you increase this inconsistency.

> 3. Is it possible to modify the number of keypoints? For instance, in my scenario, I only need four keypoints connected in the order A → B → C → D which the label is [class x y w h kpA_x kpA_y kpA_v kpB_x kpB_y kpB_v, ..., kpD_x, kpD_y, kpD_v]

You can modify number of keypoints, number classes. That's not the issue. The issue is, each keypoint should have a designated and consistent meaning across all images and augmentations.

> 4. If the number of keypoints cannot be changed, what would be the recommended approach to adapt YOLO Pose for this custom case?

The number of keypoints can be changed like I said. But the order has to be consistent across all labels. Otherwise, you will have to change the loss function so that it doesn't care about the order of the keypoints in the labels, similar to how the order of the bounding boxes are not important.

**Comment by ipmLessing** [comment]
@Y-T-G 

Thank you for the explanation!

From what I understand, in my case, it seems the best approach would be to modify the loss function to ignore the order of the keypoints, as my dataset is already labeled. Additionally, my camera captures an active object that naturally rotates (e.g., by 180 degrees).

I would appreciate your insights on the following:

1. Which loss function would be most suitable for ignoring the order of keypoints, so the model learns only the spatial locations? Since the order is not important for my use case, I am looking for a way to decouple it from the learning process.
2. How might changing the loss function impact performance? My assumption is that preserving the order could help the model learn the keypoints more effectively, but I am curious to know how significantly the change would affect accuracy.
3. Regarding hyperparameter tuning, how can I adjust them to prioritize keypoint accuracy? Currently, I have decreased the box and kobj parameters (since I only have one object and do not require the bounding box or keypoint order) and increased the pose weight during training.
4. Are there any other considerations or potential issues I might have overlooked to make this approach work effectively?

Thank you once again for your help and guidance! I truly appreciate your support.

**Comment by ipmLessing** [comment]
@Y-T-G 

- After inspecting the `PoseLoss` and `KeypointLoss` functions, I found that:
  - Introducing rotation and shift penalties into the loss only adds more penalization for rotated keypoints.
  - This results in the model learning less for a rotated keypoint, but it doesn’t resolve the rotation issue entirely.

- Consider the following scenario:
  - If the square is rotated by 5 to 10 degrees, `kp1` would still correspond to `kp1`.
  - With a camera perspective angle (e.g., 20 degrees from the left), the square would transform into a parallelogram.

- My educated guess is:
  - In predictions, there will be noticeable shifts because YOLO Pose lacks a learning mechanism to correct rotation and perspective shifts.
  - YOLO Pose seems to assume:
    - The camera is perfectly aligned.
    - The object is flat and directly facing the camera at an ideal angle.
  - This assumption makes YOLO Pose less effective in practical use, as such ideal conditions are unlikely in real-world scenarios.

**Comment by Y-T-G** [comment]
The correct loss function if you don't want it to care about order would be similar to how bounding box loss works. The order of the bounding boxes in the labels do not matter. Each GT box is assigned to an appropriate anchor based on its location and the loss is calculated based on that. To make the keypoints behave similarly, you will need to create a similar label assignment strategy for keypoints and then calculate loss based on that.
